src/analysis/constant_speed.lean

```diff
@@ -0,0 +1,224 @@
+by simp only [has_bounded_variation_on, h hx hy, ne.def, ennreal.of_real_ne_top, not_false_iff]
+
+lemma has_constant_speed_on_with_of_subsingleton
+  (f : ℝ → E) {s : set ℝ} (hs : s.subsingleton) (l : ℝ≥0) : has_constant_speed_on_with f s l :=
+begin
+  rintro x hx y hy, cases hs hx hy,
+  rw evariation_on.subsingleton f (λ y hy z hz, hs hy.1 hz.1 : (s ∩ Icc x x).subsingleton),
+  simp only [sub_self, mul_zero, ennreal.of_real_zero],
+end
+
+lemma has_constant_speed_on_with.iff_ordered :
+  has_constant_speed_on_with f s l ↔
+  ∀ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), (x ≤ y) →
+    evariation_on f (s ∩ Icc x y) = ennreal.of_real (l * (y - x)) :=
+begin
+  refine ⟨λ h x xs y ys xy, h xs ys, λ h x xs y ys, _⟩,
+  rcases le_total x y with xy|yx,
+  { exact h xs ys xy, },
+  { rw [evariation_on.subsingleton, ennreal.of_real_of_nonpos],
+    { exact mul_nonpos_of_nonneg_of_nonpos l.prop (sub_nonpos_of_le yx), },
+    { rintro z ⟨zs, xz, zy⟩ w ⟨ws, xw, wy⟩,
+      cases le_antisymm (zy.trans yx) xz,
+      cases le_antisymm (wy.trans yx) xw,
+      refl, }, },
+end
+
+lemma has_constant_speed_on_with.iff_variation_on_from_to_eq :
+  has_constant_speed_on_with f s l ↔ (has_locally_bounded_variation_on f s ∧
+  ∀ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), variation_on_from_to f s x y = l * (y - x)) :=
+begin
+  split,
+  { rintro h, refine ⟨h.has_locally_bounded_variation_on, λ x xs y ys, _⟩,
+    rw has_constant_speed_on_with.iff_ordered at h,
+    rcases le_total x y with xy|yx,
+    { rw [variation_on_from_to.eq_of_le f s xy, h xs ys xy,
+          ennreal.to_real_of_real (mul_nonneg l.prop (sub_nonneg.mpr xy))], },
+    { rw [variation_on_from_to.eq_of_ge f s yx, h ys xs yx ,
+          ennreal.to_real_of_real  (mul_nonneg l.prop (sub_nonneg.mpr yx)),
+          mul_comm ↑l, mul_comm ↑l, ←neg_mul, neg_sub], }, },
+  { rw has_constant_speed_on_with.iff_ordered,
+    rintro h x xs y ys xy,
+    rw [←h.2 xs ys, variation_on_from_to.eq_of_le f s xy,
+        ennreal.of_real_to_real (h.1 x y xs ys)], },
+end
+
+lemma has_constant_speed_on_with_zero_iff :
+  has_constant_speed_on_with f s 0 ↔ ∀ x y ∈ s, edist (f x) (f y) = 0 :=
+begin
+  dsimp [has_constant_speed_on_with],
+  simp only [zero_mul, ennreal.of_real_zero, ←evariation_on.eq_zero_iff],
+  split,
+  { by_contra',
+    obtain ⟨h, hfs⟩ := this,
+    simp_rw evariation_on.eq_zero_iff at hfs h,
+    push_neg at hfs,
+    obtain ⟨x, xs, y, ys, hxy⟩ := hfs,
+    rcases le_total x y with xy|yx,
+    { exact hxy (h xs ys x ⟨xs, le_rfl, xy⟩ y ⟨ys, xy, le_rfl⟩), },
+    { rw edist_comm at hxy,
+      exact hxy (h ys xs y ⟨ys, le_rfl, yx⟩ x ⟨xs, yx, le_rfl⟩), }, },
+  { rintro h x xs y ys,
+    refine le_antisymm _ (zero_le'),
+    rw ←h,
+    exact evariation_on.mono f (inter_subset_left s (Icc x y)), },
+end
+
+lemma has_constant_speed_on_with.ratio {l' : ℝ≥0} (hl' : l' ≠ 0) {φ : ℝ → ℝ}
+  (φm : monotone_on φ s)
+  (hfφ : has_constant_speed_on_with (f ∘ φ) s l)
+  (hf : has_constant_speed_on_with f (φ '' s) l')
```

Comment:
This one should be true if you only assume that phi maps s to t, and `hf` holds on `t`. Same thing for the unit speed version.

Reply:
Let `f : \R -> \R` be the indicator function of `2\Z` (the even numbers), let `s = 2\Z` and `t = \Z` and `\phi : \R -> \R` be the identity. Then `f` has speed `1` on `t` but speed `0` on `s` and the theorem fails, I believe. I also believe it would work if we added this condition (mentioned above), that intervals intersected with `t` are the same as intervals intersected with `\phi '' s`.

Reply:
Yes, you're right! Then the current version is perfect.